locator/tablets.cc

```diff
@@ -245,9 +246,19 @@ dht::token_range tablet_map::get_token_range(tablet_id id) const {
     }
 }
 
-host_id tablet_map::get_primary_replica(tablet_id id) const {
+tablet_replica tablet_map::get_primary_replica(tablet_id id) const {
     const auto info = get_tablet_info(id);
-    return info.replicas.at(size_t(id) % info.replicas.size()).host;
+    return info.replicas.at(size_t(id) % info.replicas.size());
```

Comment:
I am wondering if this gives enough randomness. 

Consider:

tablet0   n1, n2, n3             (n1 is the primary replica)
tablet1    n2, n1, n3             (n1 is the primary replica)
tablet2   n2, n3, n1                (n1 is the primary replica)

In this case, n1 is always the primary replica.

Reply:
It's true that the node order is shuffled in the replica set so it could result in such unfairness.
We can sort it in a predetermined order, say by host_id order and then select the primary using the ordinal tablet_id modulo number of replicas, and for the within_dc case, also filter the nodes for sorting based on the given dc.

Reply:
done in v5